### ISSUE TITLE
Evaluate PDF Structure Tag Quality and Completeness

### ISSUE THREAD

**Issue by donaldbraman** [donaldbraman/docling-testing#28]
# Issue #22: Evaluate PDF Structure Tag Quality and Completeness

## Summary
Conduct rigorous evaluation of the 90 tagged PDFs in our collection to assess the quality, completeness, and usability of their internal semantic structure tags for training data extraction. This is critical for determining whether tagged PDFs can serve as a reliable ground truth source for our 7-class document segmentation model.

## Context
Recent analysis revealed that 90 of our 427 law review PDFs (42.1%) contain internal PDF structure tags (StructTreeRoot). These tags provide semantic labels for text blocks (e.g., `<P>` for paragraphs, `<H1>` for headings, `<Note>` for footnotes) that could serve as high-quality ground truth labels for training.

However, PDF tagging completeness varies widely:
- Some PDFs may have tags for only certain elements
- Tag granularity may not match our 7-class schema
- Tag accuracy is unknown

**Ref:** docs/multiclass-training.md - Phase 1: "Rigorous Baseline Evaluation and Error Analysis"

## Objective
Evaluate whether tagged PDFs can provide sufficient ground truth data for training our multiclass document classifier. Determine:
1. **Coverage**: What percentage of content is actually tagged?
2. **Quality**: How accurately do tags represent semantic content?
3. **Completeness**: Can we map PDF tags to our 7-class schema?
4. **Usability**: Can we extract labeled training samples programmatically?

### 1. Tag Coverage Analysis
- [ ] Analyze all 90 tagged PDFs to measure what percentage of content has tags
  - Extract coverage metric: `(tagged_text_bytes / total_text_bytes) * 100`
  - Report: minimum, maximum, mean, median, std deviation
  - Breakdown by journal/source
- [ ] Document which content types are tagged vs. untagged (e.g., headers/footers may be untagged)
- [ ] Identify patterns: "PDFs from Producer X have Y% coverage"

### 2. Tag Type Inventory and Mapping
- [ ] Extract complete inventory of tag types found across all 90 PDFs
- [ ] For each tag type, determine PDF/UA standard it implements (if any)
- [ ] Map PDF standard tags to our 7-class schema:
  - `<P>` → `body_text`
  - `<H1>`, `<H2>`, `<H3>` → `heading`
  - `<Note>`, `<Footnote>` → `footnote`
  - `<Figure>`, `<Caption>` → `caption`
  - `<Artifact>`, `<Header>` → `page_header`
  - `<Artifact>`, `<Footer>` → `page_footer`
  - `<Document>` itself → `cover` (if on page 1)
- [ ] Report: What percentage of tags map cleanly to our schema?
- [ ] Document: What tags cannot be mapped and why?

### 3. Quality and Accuracy Spot Check
- [ ] **Manual validation** of 20 tagged PDFs (representative sample across journals):
  - Extract 5-10 text blocks from each PDF
  - For each block, manually verify:
    - Is the PDF tag semantically correct? (binary: yes/no)
    - Does it match the content? (e.g., is tagged `<P>` actually prose?)
    - Are any important elements missing tags?
  - Record: For each PDF, `(correct_tags / total_tagged_blocks) * 100`
- [ ] Report accuracy statistics:
  - Overall accuracy across 20 PDFs
  - Accuracy by tag type (e.g., `<P>` accuracy, `<H1>` accuracy)
  - Journals with highest/lowest accuracy
  - Any systematic bias (e.g., certain producers are consistently better/worse)

### 4. Extraction Feasibility
- [ ] Write a prototype extraction script that:
  - Reads PDF structure tree
  - Maps tagged text blocks to our 7-class labels
  - Outputs labeled samples in format: `{text, label, bbox_coords, source_pdf}`
- [ ] Test on 5 representative PDFs
- [ ] Document: What challenges arise? (e.g., tag hierarchy complexity, missing bboxes)
- [ ] Estimate: How many labeled samples can we extract from all 90 PDFs?

### 5. Comparative Analysis
- [ ] Compare quality of tagged PDFs vs. HTML-PDF pairs:
  - **Metric 1**: Completeness (% of document content with labels)
  - **Metric 2**: Granularity (how detailed are the labels?)
  - **Metric 3**: Accuracy (manual spot-check agreement)
  - **Metric 4**: Usability (ease of programmatic extraction)
- [ ] Recommendation: Which is better for training?

## Success Metrics

The evaluation is successful if we can answer:
1. **Coverage threshold**: Do >70% of PDFs have >80% content coverage? (If no, tagging is incomplete)
2. **Quality threshold**: Do >80% of sampled tags match ground truth? (If no, tags are unreliable)
3. **Mapping threshold**: Can >85% of tags be mapped to our schema? (If no, schema mismatch)
4. **Sample yield**: Can we extract >100 labeled samples per PDF on average?

If all thresholds met → **Usable for training**  
If 3/4 thresholds met → **Partially usable, requires filtering**  
If <3/4 thresholds met → **Not suitable, continue with HTML-PDF pairs**

## Deliverables

1. **PDF Tag Quality Report** (`docs/pdf_tag_quality_report.md`)
   - Coverage analysis with statistics and visualizations
   - Tag inventory and schema mapping table
   - Quality/accuracy findings with confidence intervals
   - Recommendation for use in training

2. **Extraction Script** (`scripts/corpus_building/extract_tagged_pdf_labels.py`)
   - Reads PDF structure tree
   - Maps to 7-class schema
   - Outputs CSV: `text, label, x0, y0, x1, y1, pdf_source, accuracy_confidence`
   - Handles edge cases gracefully

3. **Metadata File** (`data/tagged_pdfs_inventory.json`)
   ```json
   {
     "pdf_filename": {
       "coverage_percent": 85.3,
       "tag_types": ["P", "H1", "H2", "Note"],
       "sample_count": 142,
       "journal": "Michigan Law Review",
       "producer": "Appligent AppendPDF Pro 6.3",
       "quality_assessment": "high"
     }
   }
   ```

## Implementation Notes

- Use `pypdf` library (already in dependencies) to access StructTreeRoot
- For bounding boxes: map PDF coordinates to normalized [0-999] range (per multiclass-training.md Phase 3)
- Manual validation: use random sampling, not cherry-picking
- Report confidence intervals on all accuracy metrics (95% CI)
- Consider producer software: some tools tag better than others (e.g., Aspose vs. Prince)

## Testing

- [ ] All 90 tagged PDFs are analyzed without errors
- [ ] Coverage calculations verified on sample PDFs
- [ ] Manual spot-check results reproducible
- [ ] Extraction script tested on diverse PDFs (small, large, complex layouts)
- [ ] Report contains visualizations (coverage distribution, accuracy by journal, etc.)

## References

- docs/multiclass-training.md (Phase 1: Baseline Evaluation)
- docs/multiclass-training.md (Phase 3: Spatial Integration)
- pypdf documentation: https://pypdf.readthedocs.io/

## Labels
`data-collection`, `corpus-building`, `quality-assurance`, `7-class-schema`, `priority:high`

## Milestone
Corpus Expansion for DoclingBERT v3 (Issue #21)
